sway/desktop/output.c

```diff
@@ -601,11 +630,15 @@ static int output_repaint_timer_handler(void *data) {
 	pending.committed |= WLR_OUTPUT_STATE_DAMAGE;
 	get_frame_damage(output, &pending.damage);
 
-	if (fullscreen_con && fullscreen_con->view && !debug.noscanout) {
+	if (output_can_tear_fullscreen_view(output, fullscreen_view)) {
+		pending.tearing_page_flip = true;
```

Comment:
Asking for tearing can cause the page-flip to fail for hardware-specific reasons. Ideally we'd have some kind of fallback for that case.

Reply:
how do we know if a page has failed, and what do we want to do in in that case?

Reply:
> how do we know if a page flip has failed

The output commits fails.

> what do we want to do in in that case?

Probably fall back to a regular page-flip?

Reply:
It can also fail if you have hardware cursors enabled and a commit includes a cursor framebuffer ID change, ie. a cursor change. And this property change will persist until it succeeds. And from the kernel source, this sort of property change must be committed synchronously.

I proposed a change to wlroots, but Simon Ser said it should ideally be handled by the compositors, or at least, compositors should be aware of when flips fail and opt in to doing something different.

It may be possible to separate out the non-primary plane actions to a synchronous commit, but again, it should probably be the compositor opting into doing this.

Reply:
> It can also fail if you have hardware cursors enabled and a commit includes a cursor framebuffer ID change, ie. a cursor change. And this property change will persist until it succeeds. And from the kernel source, this sort of property change must be committed synchronously.
> 
> I proposed a change to wlroots, but Simon Ser said it should ideally be handled by the compositors, or at least, compositors should be aware of when flips fail and opt in to doing something different.
> 
> It may be possible to separate out the non-primary plane actions to a synchronous commit, but again, it should probably be the compositor opting into doing this.

Right, now I remember that was my end result of my investigation as to why atomic async page flips weren't working with this PR. Would it be possible to separate out the page flips for non-primary panes to a different synchronous commit, and if so, would that introduce any new issues?

Reply:
Just found your solution to this in your labwc AUR package and applied an equivalent solution in my tearing_control branch. Thanks for the help here!

Reply:
I intend to file it as a PR there when the wlroots 0.18 changeover happens